cpp/include/Ice/ProxyFunctions.h

```diff
@@ -0,0 +1,126 @@
+//
+// Copyright (c) ZeroC, Inc. All rights reserved.
+//
+
+#ifndef ICE_PROXY_FUNCTIONS_H
+#define ICE_PROXY_FUNCTIONS_H
+
+#include "Ice/Locator.h"
+#include "Ice/Router.h"
+
+namespace IceInternal
+{
+
+ICE_API ReferencePtr setLocator(const Ice::ObjectPrx& proxy, const std::optional<Ice::LocatorPrx>& locator);
+ICE_API ReferencePtr setRouter(const Ice::ObjectPrx& proxy, const std::optional<Ice::RouterPrx>& router);
+
+}
+
+namespace Ice
+{
+
+/**
+ * Obtains the locator configured on a proxy.
+ * @param proxy The proxy.
+ * @return The locator for the proxy. If no locator is configured, the return value is nullopt.
+*/
+ICE_API std::optional<LocatorPrx> ice_getLocator(const ObjectPrx& proxy);
+
+/**
+ * Obtains the router configured on a proxy.
+ * @param proxy The proxy.
+ * @return The router for the proxy. If no router is configured, the return value is nullopt.
+*/
+ICE_API std::optional<RouterPrx> ice_getRouter(const ObjectPrx& proxy);
+
+/**
+ * Obtains a proxy that is identical to the source proxy, except for the locator.
+ * @param proxy The source proxy.
+ * @param locator The locator for the new proxy.
+ * @return A proxy with the specified locator.
+*/
+template<typename Prx,
+         typename std::enable_if<std::is_base_of<ObjectPrx, Prx>::value>::type* = nullptr>
+Prx ice_locator(const Prx& proxy, const std::optional<LocatorPrx>& locator)
+{
+    auto ref = IceInternal::setLocator(proxy, locator);
+    return ref == proxy._getReference() ? proxy : Prx(ref);
+}
+
+/**
+ * Obtains a proxy that is identical to the source proxy, except for the router.
+ * @param proxy The source proxy.
+ * @param router The router for the new proxy.
+ * @return A proxy with the specified router.
+*/
+template<typename Prx,
+         typename std::enable_if<std::is_base_of<ObjectPrx, Prx>::value>::type* = nullptr>
+Prx ice_router(const Prx& proxy, const std::optional<RouterPrx>& router)
+{
+    auto ref = IceInternal::setRouter(proxy, router);
+    return ref == proxy._getReference() ? proxy : Prx(ref);
+}
```

Comment:
Wonder if we should pass by value, something to consider:

```
Prx ice_router(Prx proxy, const std::optional<RouterPrx>& router)
{
    auto ref = IceInternal::setRouter(proxy, router);
    return ref == proxy._getReference() ? std::move(proxy) : Prx(ref);
}

prx = ice_router(std::move(prx), router);
```

No copy in the case router is already set

Reply:
I moved back this code to ObjectPrx member functions.